keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/ImageTagger.kt

```diff
@@ -0,0 +1,90 @@
+package com.netflix.spinnaker.keel.ec2
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.module.kotlin.convertValue
+import com.netflix.spectator.api.BasicTag
+import com.netflix.spectator.api.Registry
+import com.netflix.spinnaker.keel.api.actuation.TaskLauncher
+import com.netflix.spinnaker.keel.api.plugins.CurrentImages
+import com.netflix.spinnaker.keel.core.api.DEFAULT_SERVICE_ACCOUNT
+import com.netflix.spinnaker.keel.telemetry.VerificationCompleted
+import kotlinx.coroutines.runBlocking
+import org.slf4j.LoggerFactory
+import org.springframework.context.event.EventListener
+import org.springframework.core.env.Environment
+import org.springframework.stereotype.Component
+
+/**
+ * This class tags images after they've been verified.
+ *
+ * This is meant to provide a hook to transition from managed
+ * delivery back to pipelines. Images will be tagged when they have
+ * successfully been verified with `latest tested = true`. This allows
+ * a pipeline to find the latest tested image and deploy that.
+ *
+ * This runs for all images that have passed verification.
+ */
+@Component
+class ImageTagger(
+  private val mapper: ObjectMapper,
+  private val taskLauncher: TaskLauncher,
+  private val springEnv: Environment,
+  private val spectator: Registry
+) {
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  private val TAG_AMI_JOB_LAUNCHED = "keel.image.tag"
+
+  private val shouldTagImages: Boolean
+    get() = springEnv.getProperty("keel.image.tagging.enabled", Boolean::class.java, true)
+
+  @EventListener(VerificationCompleted::class)
+  fun onVerificationCompleted(event: VerificationCompleted) {
+    if (event.status.failed() || !shouldTagImages) {
+      return
+    }
+
+    val imagesString: Any = event.metadata["images"] ?: emptyList<CurrentImages>()
```

Comment:
This code looked a little weird to me on first glance, because:

* there's a variable with the name `imagesString` that doesn't always contain a string
* I didn't know you could pass an empty list as an argument to `mapper.convertValue`

Reply:
Ya, I'll make this better. But I totally have a test for the empty images case and it works! lol

Reply:
fixed the name.